### DISCUSSION TITLE
How to notarize on MAC.

### DISCUSSION THREAD

**Discussion by raresn** [airsdk/Adobe-Runtime-Support#3075] [Guides & Tutorials]
So, in the past i had a set of commands that just worked, but now i moved on from altool to the notarytool, and i can't get the app signed.
Here is what i used to do:
**1 & 2**
AdobeAIRSDK/bin/adt -package -storetype pkcs12 -tsa none -keystore androidMarket.p12 -storepass *** -target bundle Agenda.app Agenda-Desktop.xml AgendaDesktop.swf icons images data -extdir extensions
AdobeAIRSDK/bin/adt -sign -storetype KeychainStore -alias "Developer ID Application: On Time Agenda SRL" -target bundle Agenda.app OnTimeAgenda.app
**3**
make dmg with the DROP DMG APP
**4**
xcrun altool --type osx --file OnTimeAgenda.dmg --primary-bundle-id ro.fxstudio.AgendaDesktop --notarize-app --username my_email@mail.com --asc-provider team_id
**5** 
enter this password when prompted: APP SPECIFIC PASSWORD
_pass here_
6 
If no errors uploading On Time Agenda:
xcrun stapler staple OnTimeAgenda.dmg

Now, i **changed step 4**, because altool failed to this:
xcrun notarytool submit --keychain-profile PROFILE OnTimeAgenda.dmg

Note:
The profile PROFILE was created using this:
xcrun notary tool store-credentials PROFILE --apple-id my@mail --team-id my_team_id

After running step 4, the upload is a success, but when i run:
xcrun notarytool history --keychain-profile PROFILE, my submission status is: "Invalid"
I did a xcrun notarytool log ID_OF_SUBMISSION_HERE
and i got:
statusSummary: "Archive contains critical validation errors"
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/1768694/b907bb67-a891-4592-97d4-4c64b4c21a81)

@ajwfrost or anyone who has been through this process do you have any clue what i can try next? (i've been at this for 5 hours now :) )

**Comment by ajwfrost** [comment]
Hi @raresn 

Which AIR SDK version are you using? And which macOS version?
Just wondering, if you run `java -version`, what version you're getting?

Are you able to check the output of ADT when you do the `-sign` command, from the adt.log file or the SDK Manager troubleshooting page? We should just be calling the native `codesign` process here, but perhaps something's not quite set up right..

The new notarytool steps look okay but the error is definitely suggesting a problem with the signing part. (You can probably check that with `codesign -v -vvv OnTimeAgenda.app` prior to creating the DMG file...)

thanks

**Reply by raresn** [reply]
@ajwfrost 
macOS Ventura 13.6 (22G120)
Air 50.2.3.6
codesign -v -vvv Ota.app
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/1768694/19c5c943-5dfb-46c4-95e1-0982fc5f3abf)

java -version 
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/1768694/5827fd75-760b-4ab5-962a-fea6c42685d1)

**Reply by raresn** [reply]
[adt.log](https://github.com/airsdk/Adobe-Runtime-Support/files/14336225/adt.log)
Here is the log file

**Reply by marchbold** [reply]
Have you been through this at all:

https://airsdk.dev/docs/tutorials/platform/macos/creating-macos-pkg-files

**Reply by raresn** [reply]
@marchbold thanks for the link! I looked over it just now, but that process is obsolete because it uses altool, not notarytool as required by Apple, but i guess that is not the problematic step as it can be replaced by the notarytool command. (another thing different is that it requires to create a PKG file, and we need DMG, but from what i read, that again is not really a problem). 
What @ajwfrost is trying to determine is what is going wrong in the signing process ( 'AdobeAIRSDK/bin/adt -sign -storetype KeychainStore -alias "Developer ID Application"' ). If you have any idea based on the info provided above, I would really like to try it out and see if we can get any progress.

**Reply by marchbold** [reply]
Ah yep cool, just someone added it to the site last year so thought it might have been relevant. Haven't tried it myself recently

**Reply by pol2095** [reply]
1) create a directory "Notarization"
2) sign the bundle using `adt`
3) see [here](https://blog.macadmin.me/posts/apple-notarytool/) to use `notarytool` and staple

**Reply by ajwfrost** [reply]
@raresn  this is odd because your log shows the signing process happening properly with no errors, and your codesign verification output also seems good. 

Could you perhaps just double-check it via:
```
find OnTimeAgenda.app -name "*.framework" -exec codesign -v -vvv {} \;
```
in case that first output didn't check into the framework files...?

thanks

**Reply by pol2095** [reply]
I never tried for a dmg, but for a pkg I need to notarize the bundle and after create and sign the pkg and notarize it
```
xcrun notarytool history --keychain-profile myProfile
```
return
```
Successfully received submission history.
  history
    --------------------------------------------------
    createdDate: 2024-01-20T16:22:42.325Z
    id: d-----f9-2--4-4--7-9--9-26-----17a8
    name: myApp.pkg
    status: Accepted
    --------------------------------------------------
    createdDate: 2024-01-20T16:19:39.325Z
    id: d-----f9-2--4-4--7-9--9-26-----17a8
    name: myApp.app.zip
    status: Accepted
```

**Reply by raresn** [reply]
@ajwfrost find OnTimeAgenda.app -name "*.framework" -exec codesign -v -vvv {} ;
results in:

```
OnTimeAgenda.app/Contents/Resources/META-INF/AIR/extensions/com.distriqt.Core/META-INF/ANE/MacOS-x86-64/Core.framework: code object is not signed at all
In architecture: x86_64
OnTimeAgenda.app/Contents/Resources/META-INF/AIR/extensions/com.distriqt.Core/META-INF/ANE/MacOS-x86-64/CoreNativeExtension.framework: valid on disk
OnTimeAgenda.app/Contents/Resources/META-INF/AIR/extensions/com.distriqt.Core/META-INF/ANE/MacOS-x86-64/CoreNativeExtension.framework: satisfies its Designated Requirement
OnTimeAgenda.app/Contents/Frameworks/Adobe AIR.framework: valid on disk
OnTimeAgenda.app/Contents/Frameworks/Adobe AIR.framework: satisfies its Designated Requirement

```

**Reply by raresn** [reply]
@pol2095 thanks! The notarization part i got figure out, but it failes because **there is a problem in my signing process**. Thats what we are trying to figure out. How do you sign your bundle?

**Reply by pol2095** [reply]
1) create a `Notarization` directory
`adt -sign -storetype KeyChainStore -alias "Developer ID Application: developer name (R--------S)" -target bundle ../myApp.app myApp.app`

2) `ditto -c -k myApp.app myApp.app.zip`
`xcrun notarytool submit myApp.app.zip --keychain-profile myProfile`
wait a moment...

3) `xcrun stapler staple myApp.app`

**Reply by raresn** [reply]
FINALLY! it worked, here is what i did:
1) find OnTimeAgenda.app -name "*.framework" -exec codesign -v -vvv {} \;
This shows that a framework is not signed.

2) I used codesign to sign that framework that was not signed like this:
codesign -s  "Developer ID Application: On Time Agenda SRL" --timestamp --options runtime --force OnTimeAgenda.app/Contents/Resources/META-INF/AIR/extensions/com.distriqt.Core/META-INF/ANE/MacOS-x86-64/Core.framework

3) convert to dmg

4) Submit the file again using:
xcrun notarytool submit --keychain-profile PROFILE OnTimeAgenda.dmg 

5) Doing this, i get an id, that i then use to check what happens with that submission
And now, i check that submission
With this: 
xcrun notarytool log --keychain-profile PROFILE (submission_id)

And I get a list of other frameworks/files/folders that have not been signed.
I took one by one (i had around 5), and i repeated steps 2-5 untill step 5 did not show any errors.

6) After it showed no more errors i used this command to check,
xcrun notarytool history --keychain-profile PROFILE 

7) then staple the result using:
xcrun stapler staple OnTimeAgenda.dmg

Although I was able to notarize, i find it super strange that the -sign command i ran did not sign everything.
I think this is wrong, and should sign everything inside the .app file without us going to the files one by one, right @ajwfrost?

Thank you for all the support @ajwfrost , @pol2095 , @marchbold ! #community

**Reply by ajwfrost** [reply]
> Although I was able to notarize, i find it super strange that the -sign command i ran did not sign everything.
I think this is wrong, and should sign everything inside the .app file without us going to the files one by one, right @ajwfrost?

Yes, definitely odd that this didn't work! Originally I think that `--deep` option was meant to sign a bundle and all its contents, but that was changed a while back. Hence our use of "find" which should mean that the codesign command is executed for all of these embedded framework bundles. The only thing I am wondering is whether our use of an entitlements file is causing problems... but you're not using a new macOS/Xcode version, and I believe that I'm right in thinking that some frameworks appeared to be signed and some didn't?

Will have to see if we can get some extra debugging/output into this and see what's happening in more detail... just to check, you weren't logged in via ssh or similar when you did the ADT sign command?

thanks

**Reply by raresn** [reply]
Correct, some where signed and some weren't. I wasn't logged in via ssh, and yes i am not using the new macOS/Xcode. Just a virtual mac with ventura 13.6. If i can help in any way, i am here!

**Reply by raresn** [reply]
Btw, maybe this will help you narrow things down - did a test, without any extensions, and the sign/notary processs just worked. So there is a problem when adding extensions in the xml.
